KohaRest driver implementation

EreMaijala opened the pull request:
This is a feature-complete (as far as what VuFind by default supports) driver implementation for Koha REST APIs. Currently requires Koha's master or upcoming release 20.05 and a Koha REST API plugin at https://github.com/natlibfi/koha-plugin-rest-di.

TODO:
- [x] Checkout history needs sorting by title, not yet built-in in Koha.
- [x] Laminas migration

EreMaijala committed:
Initial support for Koha REST APIs.

EreMaijala committed:
Display ILS account expiration in profile page.

EreMaijala committed:
Include 'frozen' flag for holds for extended functionality.

xmorave2 commented:
We can close #874 I think, Ere's work is great and we should base this driver on it ;) thanks!

demiankatz commented:
Great, thanks, @xmorave2; I'll take care of that now!

EreMaijala committed:
Include hold count in holding data.

demiankatz committed:
Merge branch 'master' into koharest

demiankatz committed:
Added comment.

misilot commented:
Do you know if Course Reserves will be included? As I noticed getCourses(),  getInstructors(), and getDepartments() appear to be missing?

EreMaijala committed:
Several fixes and addition of transaction history support.

EreMaijala committed:
Merge commit 'dbdba59a2f91a98309bf86aaa2de7ee24b77b2e7' into koharest

EreMaijala committed:
Update KohaRest.ini.

EreMaijala commented:
Updated to the latest version. This is a candidate for production use.

We don't use course reserves so I'm afraid that it'll be up to someone else to implement support for them.

demiankatz commented:
When you say this is a candidate for production use, do you mean that things have stabilized on the Koha side and we should consider merging this? If so, one quick suggestion: maybe we should add a note to the driver list in config.ini specifying the minimum Koha version required for the driver.

EreMaijala commented:
What I mean is that the first Finnish Koha libraries are just about to go live with Finna (VuFind), but I think the mainline Koha is still missing most of the required API endpoints. See https://github.com/Koha-Community/Koha/tree/master/api/v1/swagger/paths compared to https://github.com/KohaSuomi/Koha/tree/ks-rumble/api/v1/swagger/paths. Unfortunately it looks like very little progress has been made on Koha-Community side to integrate the features.

demiankatz commented:
Thanks for the clarification. I guess we'll leave it as a PR for now!

EreMaijala committed:
Merge branch 'master' into koharest

EreMaijala committed:
Update to latest KohaRest driver.

demiankatz committed:
Merge branch 'master' into koharest

demiankatz committed:
Style adjustment.

demiankatz commented:
@EreMaijala, I've just merged the latest master into this and resolved conflicts. Any news on the progress of the API release? I notice that the ticket cited above appears not to have changed in the past year.

EreMaijala commented:
@demiankatz Thanks! The required APIs will probably be converted to an API plugin for Koha since there's now infrastructure to support that on the Koha side. No timeframe for that at the moment, unfortunately..

demiankatz commented:
Good to know! Are there any tickets or other URLs we can watch in future for news on the progress of the plugin?

EreMaijala commented:
Not right now, but I'll make sure to link to any when they exist.

EreMaijala commented:
I'm now working on making the driver use Koha's existing REST APIs or API plugins whenever possible and adding a new API plugin for the rest of the needed functionality.

xmorave2 commented:
@EreMaijala Which enpoints are missing in Koha now do you think? I planned to start working on plugin for availability endpoint... That's the biggest issue for this drive I think...

EreMaijala commented:
@xmorave2 Availability is one, but but we also need at least the following:
- patron validation
- extensions to checkouts
- items
- checkout history

My work is still in progress, so there may be more and probably is.

xmorave2 commented:
@EreMaijala 
For checkouts history I have a path here: https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=17005
I am going to bring this proposal to Koha development meeting next week.

items should be implemented in Tomas's plugin for bibliocommons, I have it installed here: https://koha-devel-test.knihovna-uo.cz/api/v1/
And I can give you access for this testing instance if you want, we are using this instance for testing of connection Koha to Czech national library portal knihovny.cz (based on vufind).

EreMaijala commented:
@xmorave2 Unfortunately it seems to have the same design issue as normal checkouts, i.e. no support for paging the results. Especially the history can get huge, and without paging things just won't perform well.

xmorave2 commented:
@EreMaijala I've just added paging support to checkouts endpoint on Bug 17005 in Koha, we have plugin for that...

it uses two params: 
_page for number of page you want to get
_per_page is for page size, if it is not used in call, the default value from  RESTdefaultPageSize  system preference should be applied

EreMaijala committed:
Update to newest version.

EreMaijala commented:
@xmorave2 Looks like I'll never get a good chance to work on this. If you're willing to work on this, I won't mind. I just added our most recent version that still uses custom APIs for reference.

EreMaijala committed:
Merge commit '2b0d0c1c25b02f7fc29ed209944da5397263f4bb' into koharest

EreMaijala commented:
I'm finally going to continue work on this. The Koha APIs have been extended to cover at least almost all needed functionality (big thanks to @xmorave2 for his part).

EreMaijala commented:
Okay, I spoke too soon and there's still some functionality missing, but I can probably supplement it with an API plugin to get things rolling.

demiankatz commented:
Thanks, @EreMaijala, sounds like a plan! It will be good to see some movement on this again, in any case.

xmorave2 commented:
Great news @EreMaijala,
maybe we can cooperate a bit with this... We have working driver on top of almost clean Koha master...
We end up with this signature of makeRequest:

protected function makeRequest($hierarchy, $params = false, $method = 'GET')

which returns an array with keys
- data - decoded json
- headers - returned headers
- code - https status code of response

We remove the patron param, as we use Oauth2 client_credentials flow to authenticate API client.

When we needed to return also headers (when implementing pagination on getMyTransactionHistory), we decided not to add another param, but to return all of the potencially needed data - decoded json body, returned headers and http code.

The last think we need to implement is to add handling of custom request headers to makeRequest. This will be needed for getting embed objects from API, see https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=24302

What do you think about these changes? Would you like to have our (somehow production) code to try to merge it with yours?

EreMaijala commented:
@xmorave2 I'd love to have your code and would have no trouble basing my changes on your foundation. Feel free to point me to the code or send it by email (ere.maijala ät helsinki.fi).

I've started by analyzing what's missing, and I think I'll first add an API module to implement e.g. an API to fetch holdings/availability properly.

xmorave2 commented:
@EreMaijala That's no problem, first of all, there is our plugin to Koha for getting availability, it's basic one, but works:
https://github.com/xmorave2/koha-plugin-availabilityapi

Regarding the driver itself - we do not have implemented patronLogin, as we use only Shibboleth for logging patrons.

Our driver is based on your previous work and adapted to be work on clean Koha (but still with plugin for availability api)

The code is available from this branch: https://github.com/moravianlibrary/Knihovny.cz/tree/kohaDriver

It's our custom module and the driver file is here: https://github.com/moravianlibrary/Knihovny.cz/blob/kohaDriver/module/KnihovnyCz/src/KnihovnyCz/ILS/Driver/KohaRest.php

It's based on your work, but should work with newer and cleaner Koha version.

EreMaijala commented:
@xmorave2 Great stuff, thanks! I'll see what I need to add for it to be relatively feature-complete.

EreMaijala committed:
Merge branch 'master' of https://github.com/vufind-org/vufind into koharest

EreMaijala commented:
FYI, Koha REST plugin, or what I have so far, is now at https://github.com/NatLibFi/koha-plugin-rest-di. I'm now working on merging the best pieces of different VuFind driver versions.

EreMaijala committed:
Much more functionality.

EreMaijala committed:
Improved functionality using a Koha REST plugin API.

EreMaijala committed:
Merge commit '3ddcdd941fb1241bafc8e1fb551c0e311d55cac5' into koharest

EreMaijala commented:
With the latest commit I believe all the required functionality has been implemented with the help of the Koha API plugin. Still needs at least testing, and a new Koha release.

demiankatz commented:
Thanks, @EreMaijala, just let me know when it's time to merge this... and whether you have any thoughts on if/when/how to retire either of the other Koha drivers. (I imagine will still need all three for at least some period of time). Hopefully @xmorave2 can help with testing, as well.

EreMaijala committed:
Use the plugin for fetching checkout history for performance and sorting options.

EreMaijala commented:
Yes, it would be great if @xmorave2 could help with testing. My environment is all completely fake data, so testing with e.g. an anonymized real database might uncover missing changes.

xmorave2 commented:
@EreMaijala, @demiankatz Of course, I could test it on more 'real' data. I'll try to do it as soon as possible. Reading the code it is looking solid, so I think there will be no major issues.

EreMaijala commented:
@xmorave2 I'm actually expecting there to be more issues in the Koha plugin...

EreMaijala commented:
I'm still working on the plugin adding some functionality we need locally. Some minor changes will follow.

EreMaijala committed:
KohaRest fixes and improvements.

- Fix several bugs in holds handling.
- Improve the makeRequest method parameter handling.
- Fix and unify error handling.

EreMaijala committed:
Add support for retrieval of pickup locations for item-specific holds. Included for support of custom hold change functionality.

EreMaijala commented:
I think I've now completed all the forementioned changes. @xmorave2, feel free to start testing whenever it's suitable for you! :)

demiankatz commented:
Now that Laminas migration work has been merged to master, we should bring this PR up to date to match. Please let me know if you need any assistance from me.

EreMaijala committed:
Merge commit '73f90775463a548cc9a45509de2c45f255a33adf' into koharest

EreMaijala committed:
Migrate to Laminas.

EreMaijala committed:
Fix login failure handling and patron block mapping.

EreMaijala commented:
All done now. And I've briefly tested the functionality with newly released Koha 20.05. This should be good to go now!

demiankatz commented on languages/en.ini:
```diff
@@ -856,6 +861,13 @@ password_only_alphanumeric = "Numbers and letters A-Z only"
 password_only_numeric = "Numbers only"
 Passwords do not match = "Passwords do not match"
 past_days = "Past %%range%% Days"
+patron_status_address_missing = "Your address is missing."
```
Should there be Finnish/Swedish translations of these new strings?

demiankatz commented on config/vufind/config.ini:
```diff
@@ -220,6 +220,7 @@ session_name = VUFIND_SESSION
 ;   - Innovative (for INNOPAC; see also Sierra/SierraRest)
 ;   - Koha (basic database access only)
 ;   - KohaILSDI (more features via ILS-DI API)
+;   - KohaRest (more features via native API)
```
Should we say something about minimum required Koha version here? Should we consider deprecating either of the other drivers?

demiankatz committed:
Whitespace cleanup.

demiankatz committed:
More whitespace.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/CacheTrait.php:
```diff
@@ -91,7 +91,8 @@ protected function getCachedData($key)
         if (null !== $item) {
             // Return value if still valid:
-            if (time() - $item['time'] < $this->cacheLifetime) {
+            $lifetime = $item['lifetime'] ?? $this->cacheLifetime;
```
I think the changes to this file stand alone and would probably make sense as a commit outside of this PR. Do you have any objections to me cherry-picking these changes to master now?

EreMaijala replied:
No objections. Thanks!

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetRequestGroupPickupLocations.php:
```diff
@@ -53,6 +53,7 @@ public function handleRequest(Params $params)
     {
         $this->disableSessionWrites();  // avoid session write timing bug
         $id = $params->fromQuery('id');
+        $itemId = $params->fromQuery('itemId');
```
I don't see any corresponding code to pass this new value to the handler. Is it already getting passed but not being processed, or is something missing? As with the cache trait improvements, please let me know if you think this is something that should be merged independently ahead of the driver itself.

demiankatz reviewed (commented):
Thanks, @EreMaijala, it would be great to merge this soon so we can include it in release 7.0. I have a few (mostly logistical) questions, which you can see below.

EreMaijala committed:
Add Finnish and Swedish translations.

EreMaijala committed:
Revert changes to GetRequestGroupPickupLocations.

EreMaijala committed:
Add more information to config files.

EreMaijala commented:
@demiankatz I believe I've addressed your comments. I'll leave deprecation of the old drivers up to you. I think it would make sense since the old APIs don't get much attention and using database access is often convoluted, but I don't want to force KohaRest use on anyone.

demiankatz committed:
Merge branch 'master' into koharest

demiankatz committed:
Update comment.
